webapp/packages/core-blocks/src/FormControls/useInputAutocomplete.ts

```diff
@@ -0,0 +1,139 @@
+/*
+ * CloudBeaver - Cloud Database Manager
+ * Copyright (C) 2020-2024 DBeaver Corp and others
+ *
+ * Licensed under the Apache License, Version 2.0.
+ * you may not use this file except in compliance with the License.
+ */
+import { action, computed, observable } from 'mobx';
+import { type RefObject, useEffect, useMemo } from 'react';
+
+import { debounce, isNotNullDefined } from '@cloudbeaver/core-utils';
+
+import { useObservableRef } from '../useObservableRef.js';
+import { type SearchStrategy, useSearch } from '../useSearch.js';
+
+export type InputAutocompleteStrategy = 'startsWith' | 'contains' | 'fuzzy';
+
+interface InputAutocompleteOptions {
+  sourceHints: InputAutocompleteProposal[];
+  matchStrategy?: SearchStrategy;
+  predicate?: (suggestion: InputAutocompleteProposal, lastWord?: string) => boolean;
+}
+
+export interface InputAutocompleteProposal {
+  displayString: string;
+  replacementString: string;
+  icon?: string;
+  title?: string;
+}
+
+interface State {
+  replaceCurrentWord: (replacement: string) => void;
+  currentWord: string;
+  proposals: InputAutocompleteProposal[];
+}
+
+const INPUT_DELAY = 300;
+const SEARCH_FIELDS: Array<keyof InputAutocompleteProposal> = ['displayString', 'replacementString'];
+
+export const useInputAutocomplete = (
+  inputRef: RefObject<HTMLInputElement | HTMLTextAreaElement>,
+  { sourceHints, matchStrategy = 'contains', predicate }: InputAutocompleteOptions,
+): Readonly<State> => {
+  const search = useSearch({
+    sourceHints,
+    searchFields: SEARCH_FIELDS,
+    matchStrategy,
+    predicate,
+  });
+
+  const state = useObservableRef(
+    () => ({
+      isFound: false,
+      input: inputRef.current?.value as string | undefined,
+      selectionStart: inputRef.current?.selectionStart ?? null,
```

Comment:
this initial value will be incorrect because at the moment of initialization there may be no `current` object yet